### PR TITLE
Added support of multiadapters

### PR DESCRIPTION
**Pull request by mgonchar**
- Added new API to dispatcher
- Added information about adapter type to mfxPlatformInfo
- Added support of new API to sample_encode, sample_decode, sample_vpp

### CODE REVIEW

File: api/mfx_dispatch/windows/include/mfx_dispatcher.h [hunk 1]
```diff
@@ -218,4 +218,76 @@ FUNCTION_DESCRIPTION APIFunc[eVideoFuncTotal];
 
 extern const
 FUNCTION_DESCRIPTION APIAudioFunc[eAudioFuncTotal];
+
+
+#if (defined(_WIN64) || defined(_WIN32)) && (MFX_VERSION >= MFX_VERSION_NEXT)
```
Comment (dvrogozh) [inline api/mfx_dispatch/windows/include/mfx_dispatcher.h]: Why this API is made _windows_ dispatcher specific. I think it makes sense to consider declaring it on mediasdk API level otherwise it is not quite clear whether this API is official or not?

File: api/mfx_dispatch/windows/include/mfx_dispatcher.h [hunk 2]
```diff
@@ -218,4 +218,76 @@ FUNCTION_DESCRIPTION APIFunc[eVideoFuncTotal];
...
+    mfxAdapterInfo * Adapters;
+    mfxU32           NumAlloc;
+    mfxU32           NumActual;
+
+    mfxU16           reserved[4];
+};
+
+mfxStatus MFXQueryAdapters(mfxComponentInfo* input_info, mfxAdaptersInfo* adapters);
+mfxStatus MFXQueryAdaptersDecode(mfxBitstream* bitstream, mfxU32 codec_id, mfxAdaptersInfo* adapters);
+mfxStatus MFXQueryAdaptersNumber(mfxU32* num_adapters);
```
Comment (dvrogozh) [inline api/mfx_dispatch/windows/include/mfx_dispatcher.h]: This API follows approach: I have something, do you support that? That's a question not everyone bothers to actually ever ask, hence lower number of adoptions than there could be:). Alternative approach which I believe suites better is: give me what you support and I will check whether I can use it?

From the perspective of the 2nd API approach there would be 1 or 2 primary functions:
```
mfxStatus MFXQueryAdaptersNumber(mfxU32* num_adapters);
mfxStatus MFXQueryAdaptersInfo(mfxU32 num_adapters, mfxAdapterInfo* adapters);
```

After that we can have any number of helper functions which utilize the above query API and search thru it.

Mind that with this approach you immediately get opportunity to write very simple sample application which queries all adapters and reports back their caps.

Comment (mgonchar) [reply api/mfx_dispatch/windows/include/mfx_dispatcher.h]: I'm absolutely agree with you. But we have to define capabilities somehow. We need to define some general enough struct that can serve as description for supported features. It might happen that such struct would be giant and thus not easy to manipulate. And the second and most hard thing - we need to define those caps. Actually current API is extendable to support such approach also. We can add pointer to capabilities to ```mfxAdapterInfo```. So we can use it as you suggested in following way via proposed API: pass ```nullptr``` instead of requirements and attach capabilities struct to each ```mfxAdapterInfo```. Then MSDK lib will fill those structs with information about supported features for every adapter. I think we may introduce such feature later after merge of this API, because it will take some time to design proper caps struct and implementing such reporting mechanism.

File: api/mfx_dispatch/windows/include/mfx_dispatcher.h (hunk 1 above)
Comment (mgonchar) [reply api/mfx_dispatch/windows/include/mfx_dispatcher.h]: Currently we have only windows implementation, so we will have those defines for some time anyway. It is logical that those functions should be implemented on dispatcher layer (or in some separate lib operating on same layer than dispatcher), so technically it is not part of MSDK library.

Comment (onabiull) [reply api/mfx_dispatch/windows/include/mfx_dispatcher.h]: From precedent point of view, doc/mediasdk-man.pdf already define API in terms of lib+dispatcher. I.e. MFXInit would behave slightly differently w/o dispatcher. Absence of Linux implementation is a clear gap.

Comment (mgonchar) [reply api/mfx_dispatch/windows/include/mfx_dispatcher.h]: exposed to MSDK API

File: api/include/mfxstructures.h [hunk 3]
```diff
@@ -2269,6 +2269,48 @@ MFX_PACK_END()
 
 #endif
 
+#if (MFX_VERSION >= MFX_VERSION_NEXT)
+/* Multi-adapters Querying structs */
+typedef enum
+{
+    MFX_ENCODE_COMPONENT = 1,
```
Comment (dvrogozh) [inline api/include/mfxstructures.h]: From my perspective `MFX_COMPONENT_<name>` style will be better.

File: api/include/mfxstructures.h [hunk 4]
```diff
@@ -2269,6 +2269,48 @@ MFX_PACK_END()
...
+    MFX_DECODE_COMPONENT = 2,
+    MFX_VPP_COMPONENT    = 3
+} mfxComponentType;
```
Comment (dvrogozh) [inline api/include/mfxstructures.h]: how FEI is classified?

File: api/include/mfxstructures.h [hunk 5]
```diff
@@ -2269,6 +2269,48 @@ MFX_PACK_END()
...
+
+MFX_PACK_BEGIN_STRUCT_W_PTR()
+typedef struct
+{
+    mfxComponentType Type;
```
Comment (dvrogozh) [inline api/include/mfxstructures.h]: hm... We have few reserved fields in mfxVideoParam. Maybe we will just add `mfxComponentType` in there and avoid defining mfxComponentInfo altogether?

File: api/include/mfxstructures.h [hunk 6]
```diff
@@ -2269,6 +2269,48 @@ MFX_PACK_END()
...
+MFX_PACK_END()
+
+/* Adapter description */
+MFX_PACK_BEGIN_USUAL_STRUCT()
+typedef struct
+{
+    mfxPlatform Platform;
+    mfxU32      Number;
+
+    mfxU16      reserved[14];
```
Comment (dvrogozh) [inline api/include/mfxstructures.h]: we might need adapter type on Linux to distinguish render nodes and cards

File: api/mfx_dispatch/windows/src/mfx_dispatcher.cpp [hunk 7]
```diff
@@ -318,3 +325,292 @@ mfxStatus MFX_DISP_HANDLE::UnLoadSelectedDLL(void)
     return mfxRes;
 
 } // mfxStatus MFX_DISP_HANDLE::UnLoadSelectedDLL(void)
+
+#if (defined(_WIN64) || defined(_WIN32)) && (MFX_VERSION >= MFX_VERSION_NEXT)
+static mfxStatus InitDummySession(mfxU32 adapter_n, MFXVideoSession & dummy_session)
```
Comment (dvrogozh) [inline api/mfx_dispatch/windows/src/mfx_dispatcher.cpp]: errgghh.... I do NOT like creating any dummy sessions in the dispatcher for the reason that they are not cheap. They create scheduler and fully qualified core. If that's needed to support adapter query via bitstream I would better sacrifice this function for now.

File: api/include/mfxadapter.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "mfxstructures.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+mfxStatus MFXQueryAdapters(mfxComponentInfo* input_info, mfxAdaptersInfo* adapters);
+mfxStatus MFXQueryAdaptersDecode(mfxBitstream* bitstream, mfxU32 codec_id, mfxAdaptersInfo* adapters);
+mfxStatus MFXQueryAdaptersNumber(mfxU32* num_adapters);
```
Comment (dvrogozh) [inline api/include/mfxadapter.h]: Per my opinion this API is over complex. For example, 2 structs with almost the same name (mfxAdapterSInfo) confuse. I would this:
```
mfxU32 MFXQueryAdaptersNumber();

// num_adapters size of the adapters array allocated by the user
// functions return number of match adapters within adapters array
mfxU32 MFXQueryAdapters(mfxComponentInfo* input_info, mfxAdapterInfo* adapters, mfxU32 num_adapters);
mfxU32 MFXQueryAdaptersDecode(mfxBitstream* bitstream, mfxU32 codec_id, mfxAdapterInfo* adapters, mfxU32 num_adapters);
```

File: api/mfx_dispatch/windows/src/mfx_dispatcher.cpp (hunk 7 above)
Comment (dvrogozh) [reply api/mfx_dispatch/windows/src/mfx_dispatcher.cpp]: To solve that I would suggest to implement MFXQueryAdapters* functions directly inside mediasdk and teach dispatcher to collect this info directly from msdk libs and return aggregated info.

**Comment by mgonchar** [comment]
@dvrogozh documentation added

File: api/include/mfxstructures.h (hunk 3 above)
Comment (mgonchar) [reply api/include/mfxstructures.h]: Well, maybe

File: api/include/mfxstructures.h (hunk 4 above)
Comment (mgonchar) [reply api/include/mfxstructures.h]: FEI is ENCODE + extension buffer (attached to ```mfxVideoParam``` on Init). On HEVC currently it is separate plugin, but it is possible to initialize it same way (with ext buffer) as AVC. We just need to add such buffer to API for HEVC.

Comment (mgonchar) [reply api/include/mfxstructures.h]: But for HEVC FEI scope of supported configurations is absolutely same as for regular HEVC encoder, so it is not critical issue.

File: api/include/mfxstructures.h (hunk 6 above)
Comment (mgonchar) [reply api/include/mfxstructures.h]: Yep, those reserved fields might me reused for Linux support

File: api/mfx_dispatch/windows/src/mfx_dispatcher.cpp (hunk 7 above)
Comment (mgonchar) [reply api/mfx_dispatch/windows/src/mfx_dispatcher.cpp]: Dummy sessions used in each function. No difference for bitstream in this

File: api/include/mfxstructures.h (hunk 5 above)
Comment (mgonchar) [reply api/include/mfxstructures.h]: No, I don't like this idea. We widely use ```mfxVideoParam``` struct in other parts of API and design of this struct assumes that it is component agnostic

File: api/include/mfxadapter.h (hunk 8 above)
Comment (mgonchar) [reply api/include/mfxadapter.h]: But we also report back internal statuses, which is aligned with MSDK API, where we always return statuses

File: doc/mediasdk-man.md [hunk 9]
```diff
@@ -5671,11 +5672,14 @@ The `mfxPlatform` structure contains information about hardware platform.
 --- | ---
 `CodeName` | Intel® microarchitecture code name. See the [PlatformCodeName](#PlatformCodeName) enumerator for a list of possible values.
 `DeviceId` | Reserved.
+`MediaAdapterType` | Description of Intel Gen Graphics adapter type. See the [mfxMediaAdapterType](#mfxMediaAdapterType) enumerator for a list of possible values.
 
 **Change History**
 
 This structure is available since SDK API 1.19.
 
+The SDK API **TBD** adds `MediaAdapterType` field.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: @onabiull : how we will handle these in releases if API won't land in the release? I would suggest to make API changes in separate commits which we can specially mark somehow and just revert commits in release branch.

File: api/include/mfxadapter.h (hunk 8 above)
Comment (dvrogozh) [reply api/include/mfxadapter.h]: This is not an argument to me. Imho, "we always did that" is not kind of argument I would accept. I still am persistent with my opinion: API is over complex.

File: api/include/mfxstructures.h (hunk 5 above)
Comment (dvrogozh) [reply api/include/mfxstructures.h]: That's another badly designed API. mfxVideoParam has union inside _without_ any way to understand which union path is to be followed from the data stored in the structure. IMHO, mfxComponentType should have been initially in this union from the msdk birth day!

File: doc/mediasdk-man.md [hunk 10]
```diff
@@ -1161,6 +1165,125 @@ Assume the pipeline is **A->B->C**. The application synchronizes on sync point *
 
 ## <a id='hardware_acceleration'>Working with hardware acceleration</a>
 
+### <a id='Working_with_multiple_intel_media_devices'>Working with multiple Intel media devices</a>
+
+If your system has multiple Intel Gen Graphics adapters you may need hints on which adapter suits better to process some particular workload. The SDK provides API to help with adapter selection. See Example 9 that showcases workload processing initialization on discrete adapter.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: hard to read....

"Media SDK provides helper API to select best suitable adapter for your workload based on passed workload description. See Example 9 that showcases workload initialization on discrete adapter." (remove word "processing" from example line)

File: doc/mediasdk-man.md [hunk 11]
```diff
@@ -1161,6 +1165,125 @@ Assume the pipeline is **A->B->C**. The application synchronizes on sync point *
...
+if (idx_d == adapters.Adapters + adapters.NumActual)
+{
+    printf("Warning: No dGfx detected on machine\n");
+    return -1;
+}
+
+mfxU32 idx = static_cast<mfxU32>(std::distance(adapters.Adapters, idx));
+
+// Choose correct implementation for discrete adapter
+switch (adapters.Adapters[idx].Number)
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: hm. And why API does not return impl directly:)?

File: doc/mediasdk-man.md [hunk 12]
```diff
@@ -1161,6 +1165,125 @@ Assume the pipeline is **A->B->C**. The application synchronizes on sync point *
...
+    impl = MFX_IMPL_HARDWARE_ANY;
+    break;
+}
+
+// Initialize mfxSession in regular way with obtained implementation
+```
+
+As you see in Example 9, after obtaining adapter list with [MFXQueryAdapters](#MFXQueryAdapters) further initialization of [mfxSession](#mfxSession) is performed in regular way. Particular adapter choice is performed with [MFX_IMPL_HARDWARE](#mfxIMPL)**...**[MFX_IMPL_HARDWARE4](#mfxIMPL) values of [mfxIMPL](#mfxIMPL) (see [Multiple Monitors](#Multiple_monitors) section for details).
+
+Example 10 showcases usage of [MFXQueryAdapters](#MFXQueryAdapters) for querying best suitable adapter for particular encode workload processing (see [MFXQueryAdapters](#MFXQueryAdapters) description for adapter priority rules).
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: remove "processing"

File: doc/mediasdk-man.md [hunk 13]
```diff
@@ -1469,6 +1592,97 @@ Global functions initialize and de-initialize the SDK library and perform query
...
+
+**Syntax**
+
+[mfxStatus](#mfxStatus) `MFXQueryAdapters(mfxComponentInfo* input_info, mfxAdaptersInfo* adapters);`
+
+**Parameters**
+
+| | |
+--- | ---
+`input_info` | Pointer to workload description. See [mfxComponentInfo](#mfxComponentInfo) description for details.
```
Comment (dvrogozh) [inline doc/mediasdk-man.md]: Can I pass input_info=nullptr? What I will get?

File: api/mfx_dispatch/windows/src/mfx_dispatcher.cpp (hunk 7 above)
Comment (dvrogozh) [reply api/mfx_dispatch/windows/src/mfx_dispatcher.cpp]: There will be difference for Linux dispatcher which is tiny and I don't want to make it bigger.

File: samples/sample_decode/src/pipeline_decode.cpp [hunk 14]
```diff
@@ -141,6 +141,142 @@ CDecodingPipeline::~CDecodingPipeline()
     Close();
 }
 
+#if (defined(_WIN64) || defined(_WIN32)) && (MFX_VERSION >= MFX_VERSION_NEXT)
+mfxU32 CDecodingPipeline::GetPreferredAdapterNum(const mfxAdaptersInfo & adapters, const sInputParams & params)
```
Comment (dvrogozh) [inline samples/sample_decode/src/pipeline_decode.cpp]: Why this function is in decoder sample? Why not in utilities?

Comment (dvrogozh) [reply samples/sample_decode/src/pipeline_decode.cpp]: Do we actually need WIN protection here and in most other places? API is common between Windows and Linux...

File: samples/sample_decode/src/sample_decode.cpp [hunk 15]
```diff
@@ -62,6 +62,10 @@ void PrintHelp(msdk_char *strAppName, const msdk_char *strErrorMessage)
     msdk_printf(MSDK_STRING("   [-di bob/adi]             - enable deinterlacing BOB/ADI\n"));
 #if (MFX_VERSION >= 1025)
     msdk_printf(MSDK_STRING("   [-d]                      - enable decode error report\n"));
+#endif
+#if (defined(_WIN64) || defined(_WIN32)) && (MFX_VERSION >= MFX_VERSION_NEXT)
```
Comment (dvrogozh) [inline samples/sample_decode/src/sample_decode.cpp]: this

File: samples/sample_decode/src/sample_decode.cpp [hunk 16]
```diff
@@ -416,6 +419,16 @@ mfxStatus ParseInputString(msdk_char* strInput[], mfxU8 nArgNum, sInputParams* p
         {
             pParams->bSoftRobustFlag = true;
         }
+#if (defined(_WIN64) || defined(_WIN32)) && (MFX_VERSION >= MFX_VERSION_NEXT)
```
Comment (dvrogozh) [inline samples/sample_decode/src/sample_decode.cpp]: and this are likely the only 2 places which need WIN protection since on Linux we don't have API implemented (yet)

File: samples/sample_encode/src/pipeline_encode.cpp [hunk 17]
```diff
@@ -1350,12 +1352,181 @@ mfxStatus CEncodingPipeline::InitFileWriters(sInputParams *pParams)
...
+    case MFX_FOURCC_Y416:
+        return 12;
+        break;
+#endif
+    default:
+        return 0;
+    }
+}
+
+mfxU32 CEncodingPipeline::GetPreferredAdapterNum(const mfxAdaptersInfo & adapters, const sInputParams & params)
```
Comment (dvrogozh) [inline samples/sample_encode/src/pipeline_encode.cpp]: "he's once again my guest..."

File: samples/sample_vpp/src/sample_vpp_utils.cpp [hunk 18]
```diff
@@ -414,6 +426,107 @@ mfxStatus InitParamsVPP(mfxVideoParam* pParams, sInputParams* pInParams, mfxU32
 
 /* ******************************************************************* */
 
+#if (defined(_WIN64) || defined(_WIN32)) && (MFX_VERSION >= MFX_VERSION_NEXT)
+mfxU32 GetPreferredAdapterNum(const mfxAdaptersInfo & adapters, const sInputParams & params)
```
Comment (dvrogozh) [inline samples/sample_vpp/src/sample_vpp_utils.cpp]: emmmm.....

File: doc/mediasdk-man.md (hunk 9 above)
Comment (onabiull) [reply doc/mediasdk-man.md]: My plan is to strip all \*\*TBD\*\* sections from documents in release branch. Probably in the same commit with version change. After all there are just dozen or so of them.

File: api/include/mfxstructures.h (hunk 5 above)
Comment (mgonchar) [reply api/include/mfxstructures.h]: Well, maybe, but MSDK uses different approach for component type resolution. We use different functions calls, i.e. resolution happens on level of function name

File: api/include/mfxadapter.h (hunk 8 above)
Comment (mgonchar) [reply api/include/mfxadapter.h]: Well, from other side your approach hiding errors. What if something will go wrong? You suggest to return 0 in such case (no suitable adapters), but how user will know the reason of this? Either it is workload that can't be handled, or some internal error occured

File: api/mfx_dispatch/windows/src/mfx_dispatcher.cpp (hunk 7 above)
Comment (mgonchar) [reply api/mfx_dispatch/windows/src/mfx_dispatcher.cpp]: About implementing it in lib. How do you suppose to retrieve information from adapter? This information becomes available at init time. How should we redesign MSDK to connect unitialized library to adapter? Currently dispatcher loads lib during first init call (that's why there are dummy sessions and inits here), but how to bypass them, I don't see an easy way now, which will be better that proposed solution

File: doc/mediasdk-man.md (hunk 9 above)
Comment (mgonchar) [reply doc/mediasdk-man.md]: So, manual editing of docs?) Shouldn't we hide hints for automatic cut inside md comments for TBD sections?

File: doc/mediasdk-man.md (hunk 11 above)
Comment (mgonchar) [reply doc/mediasdk-man.md]: Because impl type supports up to 4 adapters, in future we will use ext buffer to support more adapters

File: doc/mediasdk-man.md (hunk 10 above)
Comment (mgonchar) [reply doc/mediasdk-man.md]: ok, will do later

File: doc/mediasdk-man.md (hunk 13 above)
Comment (mgonchar) [reply doc/mediasdk-man.md]: list of adapters. If I forgot to mention it in doc, I will add

File: samples/sample_decode/src/pipeline_decode.cpp (hunk 14 above)
Comment (mgonchar) [reply samples/sample_decode/src/pipeline_decode.cpp]: since it is implementation, yes we need. I can think about putting it to utility

File: samples/sample_vpp/src/sample_vpp_utils.cpp (hunk 18 above)
Comment (mgonchar) [reply samples/sample_vpp/src/sample_vpp_utils.cpp]: these are different input parameters, but they have same name. We shouldn't bring this to utility, since we don't want to put giant structs of input parameters description to common code

File: samples/sample_encode/src/pipeline_encode.cpp (hunk 17 above)
Comment (mgonchar) [reply samples/sample_encode/src/pipeline_encode.cpp]: "It's not what it seems to be..."

File: samples/sample_decode/src/sample_decode.cpp (hunk 16 above)
Comment (mgonchar) [reply samples/sample_decode/src/sample_decode.cpp]: No, because setting these two parameters to ```false``` doesn't mean bypassing new API calls. By default we will query best suitable adapter

File: samples/sample_vpp/src/sample_vpp_utils.cpp (hunk 18 above)
Comment (dvrogozh) [reply samples/sample_vpp/src/sample_vpp_utils.cpp]: I got impression that you use just 1, maybe 2 values from input parameters. If so, you can change function prototype and pass only what you need making it possible to move function to utilities library.

Comment (mgonchar) [reply samples/sample_vpp/src/sample_vpp_utils.cpp]: Not sure if it makes sense. In this case it would be something like:
```
common_header.h
...
mfxU32 GetPreferredAdapterNum(const mfxAdaptersInfo & adapters, const mfxVideoParam& params)
```

But in each sample:
```
sample.cpp
...
void FillParameters(const sInputParams & input_params, mfxVideoParam& output_params)
```

Doesn't look much better

File: doc/mediasdk-man.md (hunk 9 above)
Comment (onabiull) [reply doc/mediasdk-man.md]: Right now - yes. We can think about employing html-style comments (they won't be rendered), but I not sure that in current scheme - 4 releases per year, it would worth investments.

File: doc/mediasdk-man.md (hunk 10 above)
Comment (mgonchar) [reply doc/mediasdk-man.md]: done

File: doc/mediasdk-man.md (hunk 12 above)
Comment (mgonchar) [reply doc/mediasdk-man.md]: done

File: doc/mediasdk-man.md (hunk 13 above)
Comment (mgonchar) [reply doc/mediasdk-man.md]: Doc updated

**Comment by onabiull** [comment]
As we discussed offline, looks like accepting this in current state, are best option. Still there are clearly much room for improvement :)
